nestedtensor/nested/masking.py

```diff
@@ -25,45 +22,41 @@ def nested_tensor_from_tensor_mask(tensor, mask, nested_dim=None):
     if not mask.dim() > 0:
         raise RuntimeError("Mask has to have dimention > 0")
 
-    # TODO: Need to define empty-list and 0 numel semantics
-    # TODO: Test edge-cases and kwargs
-    def _merge(tensors, nested_dim=None):
-        assert len(tensors) > 0
+    if nested_dim is not None and mask.dim() < nested_dim:
+        raise ValueError("Mask dimension ({0}) is too small to construct nested tensor with nested dimension of ({1})".format(mask.dim(), nested_dim))
+
+    def create_nested_tensor(tensors, nested_dim=None):
+        assert len(tensors) > 0, "Can't merge empty list of tensors. Please, report this error."
 
-        def _mergable_tensors(tensors, nested_dim):
+        def should_create_nested_tensor(tensors, nested_dim):
             if nested_dim is None:
-                size0 = tensors[0].size()
-                compliant = all(torch.is_tensor(t) for t in tensors)
-                compliant = all(size0 == t.size()
-                                for t in tensors) and compliant
-                return compliant
+                size = tensors[0].size()
+                return not all(torch.is_tensor(t) and size == t.size() for t in tensors)
             else:
-                return nested_dim < 1
+                return not nested_dim < 1
 
-        if _mergable_tensors(tensors, nested_dim):
-            return torch.stack(tensors)
-        else:
+        if should_create_nested_tensor(tensors, nested_dim):
             return creation.nested_tensor(tensors)
-
-    if nested_dim is not None and mask.dim() < nested_dim:
-        raise ValueError("Mask dimension ({0}) is too small to construct nested tensor with nested dimension of {1}".format(mask.dim(), nested_dim))
+        else:
+            return torch.stack(tensors)
 
     if mask.dim() == 1:
-        tensors = [tensor[i] if mask[i]
-                   else None for i in range(len(mask))]
-        tensors = list(filter(lambda x: x is not None, tensors))
+        tensors = []
+        for i in range(len(mask)):
+            if mask[i]:
+                tensors.append(tensor[i])
+
         if len(tensors) == 0:
             return tensor.narrow(0, 0, 0)
```

Comment:
The idea here is to return an empty tensor of the same type etc. and also support autograd.